### ISSUE TITLE
E-signatures stopped working - not saving - no errors

### ISSUE THREAD

**Issue by krobertson71** [susom/redcap-em-shibboleth-esignature#5]
We are having another issue where e-signatures are no longer saving.  We cannot identify a root cause.  There are no errors in any Apache/Shibd/PHP-FPM/MySQl log to explain why this is happening. 

We are running version 1.0.3 of the module with a fix from @nsimbi from this issue https://github.com/susom/redcap-em-shibboleth-esignature/issues/2.

We cannot upgrade due to the version limitation set in 1.0.4 and above.  1.0.4 was a fix for the above issue I linked but we experienced it in 12.4.12 (and in 12.4.31 after upgrade).  We are now having a new issue which is described below.

I posted this on the REDCap community site but I am afraid they are going to blame it on the module, as in the past, and send me here anyway so I am just jumping ahead :).  

Basically we are pulling up a record, making entries, selecting lock and sign.. we get the forced authentication from our IdP,  we are taken back to the popup window that says we have been verified and click sign.. then we are taken back to the document (Say & Stay).  Normally there are two notification boxes at the top of the record, one for locking and one for e-signature, stating that they were successful.  At the bottom each check-mark will normally show an id and date/time stamp for each. Now the document just states "edited" with no locking and e-signature notifications, and no id and date/time stamp and boxes are back to being unchecked.

We checked the database table redcap_esignatures where a successful e-sig save is recorded.  Now there is no entries for any attempt made.

We did have to recover the database server (Azure MySQL Flexible Server, v8).  Originally I thought this might be the root cause but we had a UAT tester test e-signature successfully 3 hours after the recovery.  We recovered the database up to 1.5 hours before last use.

 

So we are sort of hitting a brick wall.  Any help or ideas would be greatly appreciated.

**Comment by krobertson71** [comment]
Update - I installed emLogger and set that up.  I now see these errors. As you can see below its failing on the timer thinking I took 70 and 71 seconds to sign.  I timed this myself from the time clicking "Add/Edit" records to locking and signing, authentication with shib, to back to form.... 38 seconds.. I am not sure when the timer starts but I believe it's when you click 'Sign' so I captured more than that.  I scrubbed out my internal email address.

{"date":"2023-04-01 20:08:04","type":"INFO","pid":"15","username":[blah@my.edu](mailto:blah@my.edu),"args":[{"type":"string","value":"70 have elapsed since [blah@my.edu's](mailto:blah@my.edu's) verification - not counting as valid"}],"file":"\/var\/www\/html\/redcap\/modules\/shibboleth_esignature_v1.0.3\/Esignature.php","line":78,"function":"redcap_every_page_before_render","runtime":14.1}
{"date":"2023-04-01 20:11:15","type":"INFO","pid":"33","username":[blah@my.edu](mailto:blah@my.edu),"args":[{"type":"string","value":"71 have elapsed since [blah@my.edu](mailto:blah@my.eduu's) verification - not counting as valid"}],"file":"\/var\/www\/html\/redcap\/modules\/shibboleth_esignature_v1.0.3\/Esignature.php","line":78,"function":"redcap_every_page_before_render","runtime":7}

**Comment by krobertson71** [comment]
Update #2 

I changed lifetime and session in my shibboleth2.xml to 80 seconds but still received the below error for 70 seconds- emailed scrubbed

{"date":"2023-04-01 20:58:52","type":"INFO","pid":"15","username":"me@you.com","args":[{"type":"string","value":"70 have elapsed since me@you.com's verification - not counting as valid"}],"file":"\/var\/www\/html\/redcap\/modules\/shibboleth_esignature_v1.0.3\/Esignature.php","line":78,"function":"redcap_every_page_before_render","runtime":6.1}

**Comment by 123andy** [comment]
Hi @krobertson71 - I haven't looked at this in a long while.  And, I'm currently out of the office until April 11th.  If I am able to make time during my vacation I'll look into it.    Can you tell me your REDCap version number and php version number?  My guess is:

1) Perhaps the locking code has been updated in REDCap and this is causing a problem (as the module had to copy a lot of REDCap code as there were not any official methods for this, or
2) It might be that tightening of CSRF tokens has led the ajax call to verify to fail.  

One other thing that might be helpful is to open your browser, enable developer tools, click on the network tab, click on 'preserve log' and then go through the process.  I would like to see if the opening of the verify page works and if you get any HTML errors on the client, such as CSRF...  

-A

**Comment by krobertson71** [comment]
Sorry I forgot to include all the info- 

REDCap 12.4.31 · PHP 8.0.27 (Linux/Unix OS) · MySQL 8.0.32

Also, the environment where this started failing is identical to our dev environment where e-signature is still working.  This out UAT environment.  Like I mentioned this environment had to have it's database recovered from an accidental deletion.  Granted there was a successful e-sig test 3 hours later but I still wonder if something in the db might have  gotten corrputed (wild guess).

I found the code for the checks in Esignature.php.  I see the 60 sec time is hardcoded there.  I changed the timers to 80 in that file and e-signature is working again.  Why extra time is being added to the counter in the log output (recording 71 sec vs 35 actual) eludes me though.  This is just a workaround as they like the timer at 60 seconds.

I will take a closer look at the developer tools but last pass everything was successful.  The php page itself gives no error and actually says the esignature was successful even though emLogger says it was not vaild.

**Comment by krobertson71** [comment]
We have solved the issue.  Turned out the time on the Azure VM was drifting and padding the esignature timestamp.  In your Esignature.php code the initial timestamp is recorded in the MySQL database.  So it uses the MYSql server time (which was correct) for the timestamp.  Once we complete the process you grab the current date/time (now) from the app server, which was 1 min 45 secs in the future.  Once we resolved with chronyd all was well.  Sorry to have bothered you on your vacation and hope you have a great time!
